### PR TITLE
Perform thread check only when store is created on main queue.

### PR DESCRIPTION
**Pull request by mbrandonw**
This relaxes the thread checking in the store to only those stores created on the main queue. This still allows people to create stores on background threads and send actions on background threads, but such stores will not get thread checking.

See https://github.com/pointfreeco/swift-composable-architecture/discussions/802 for a lot more discussion on why we want to do this.

### CODE REVIEW

File: Sources/ComposableArchitecture/Store.swift [hunk 1]
```diff
@@ -368,8 +373,8 @@ public final class Store<State, Action> {
     self.publisherScope(state: toLocalState, action: { $0 })
   }
 
-  func send(_ action: Action, isFromViewStore: Bool = true) {
-    self.threadCheck(status: .send(action, isFromViewStore: isFromViewStore))
+  func send(_ action: Action, originatingFrom originatingAction: Action?) {
+    self.threadCheck(status: .send(action, originatingAction: originatingAction))
```
Comment (mbrandonw) [inline Sources/ComposableArchitecture/Store.swift:377]: This is unrelated to the purpose of this PR, but by threading the originating action for effect emissions we can provide better messaging when the thread check fails.

File: Sources/ComposableArchitecture/Store.swift [hunk 2]
```diff
@@ -418,62 +423,63 @@ public final class Store<State, Action> {
...
+          called on the main thread.
           """
 
-      case let .send(action, isFromViewStore: true):
+      case let .send(action, originatingAction: nil):
         message = """
-          "ViewStore.send(\(debugCaseOutput(action)))" was called on the wrong thread. Make sure \
-          that "ViewStore.send" is always called on the same thread the store was created on.
+          "ViewStore.send(\(debugCaseOutput(action)))" was called on a non-main thread. Make sure \
+          that "ViewStore.send" is always called on the main thread..
```
Comment (stephencelis) [inline Sources/ComposableArchitecture/Store.swift:453]:
```suggestion
          that "ViewStore.send" is always called on the main thread.
```

**Review by stephencelis — APPROVED** [review]

File: Sources/ComposableArchitecture/Store.swift [hunk 3]
```diff
@@ -368,8 +373,8 @@ public final class Store<State, Action> {
     self.publisherScope(state: toLocalState, action: { $0 })
   }
 
-  func send(_ action: Action, isFromViewStore: Bool = true) {
-    self.threadCheck(status: .send(action, isFromViewStore: isFromViewStore))
+  func send(_ action: Action, originatingFrom originatingAction: Action?) {
```
Comment (stephencelis) [inline Sources/ComposableArchitecture/Store.swift:376]:
```suggestion
  func send(_ action: Action, originatingFrom originatingAction: Action? = nil) {
```

**Review by mluisbrown — APPROVED** [review]

File: Sources/ComposableArchitecture/Store.swift [hunk 4]
```diff
@@ -111,6 +111,21 @@ import Foundation
...
+/// #### Thread safety checks
+///
+/// The store performs some basic thread safety checks in order to help catch mistakes. Stores
+/// constructed via the initializer ``Store/init(initialState:reducer:environment:)`` are assumed
+/// to run only on the main thread, and so a check is executed immediately to make sure that is the
+/// case. Further, all actions sent to the store and all scopes (see ``Store/scope(state:action:)``)
+/// of the store are also checked to make sure that work is performed on the main thread.
+///
+/// If you need a store that runs on a non-main thread, which should be very rare and you should
+/// have a very good reason to do so, then you can consturct a store via the
```
Comment (tgrapperon) [inline Sources/ComposableArchitecture/Store.swift:123]: There's a typo at "construct"

File: Tests/ComposableArchitectureTests/StoreTests.swift [hunk 5]
```diff
@@ -447,4 +447,26 @@ final class StoreTests: XCTestCase {
...
+        viewStore.send(())
+        DispatchQueue.global().asyncAfter(deadline: .now() + 0.1) {
+          viewStore.send(())
+        }
+      }
+    }
+
+    // Allow asychnronous code to run for a bit to make sure the warning breakpoint is
+    // never tripped.
+    _ = XCTWaiter.wait(for: [.init()], timeout: 1)
```
Comment (tgrapperon) [inline Tests/ComposableArchitectureTests/StoreTests.swift:470]: I think we are always waiting 1s and failing silently here. Is it the expected behavior?

Comment (tgrapperon) [reply Tests/ComposableArchitectureTests/StoreTests.swift:470]: What about:
```swift
func testNonMainQueueStore() {
  var expectations: [XCTestExpectation] = []
  for i in 1 ... 100 {
    let expectation = XCTestExpectation(description: "\(i)th iteration is complete")
    expectations.append(expectation)
    DispatchQueue.global().async {
      let viewStore = ViewStore(
        Store.unchecked(
          initialState: 0,
          reducer: Reducer<Int, Void, Void> {
            state, _, _ in
            state += 1
            if state == 2 { expectation.fulfill() }
            return .none
          },
          environment: ()
        )
      )
      viewStore.send(())
      DispatchQueue.global().asyncAfter(deadline: .now() + 0.1) {
        viewStore.send(())
      }
    }
  }

  wait(for: expectations, timeout: 1)
}
```
On my machine, it runs the test in 0.13s.

Comment (mbrandonw) [reply Tests/ComposableArchitectureTests/StoreTests.swift]: Great suggestion, thanks!

Comment (tgrapperon) [reply Tests/ComposableArchitectureTests/StoreTests.swift]: In a true TCA spirit, I now think that the reducer's environment should be the expectation!

File: Tests/ComposableArchitectureTests/StoreTests.swift [hunk 6]
```diff
@@ -447,4 +447,34 @@ final class StoreTests: XCTestCase {
...
+  func testNonMainQueueStore() {
+    var expectations: [XCTestExpectation] = []
+    for i in 1 ... 100 {
+      let expectation = XCTestExpectation(description: "\(i)th iteration is complete")
+      expectations.append(expectation)
+      DispatchQueue.global().async {
+        let viewStore = ViewStore(
+          Store.unchecked(
+            initialState: 0,
+            reducer: Reducer<Int, Void, Void> {
```
Comment (tgrapperon) [inline Tests/ComposableArchitectureTests/StoreTests.swift:460]:
```suggestion
            reducer: Reducer<Int, Void, XCTestExpectation> {
```

File: Tests/ComposableArchitectureTests/StoreTests.swift [hunk 7]
```diff
@@ -447,4 +447,34 @@ final class StoreTests: XCTestCase {
...
+              state, _, _ in
```
Comment (tgrapperon) [inline Tests/ComposableArchitectureTests/StoreTests.swift:461]:
```suggestion
              state, _, expectation in
```

File: Tests/ComposableArchitectureTests/StoreTests.swift [hunk 8]
```diff
@@ -447,4 +447,34 @@ final class StoreTests: XCTestCase {
...
+              state += 1
+              if state == 2 {
+                return .fireAndForget { expectation.fulfill() }
+              }
+              return .none
+            },
+            environment: ()
```
Comment (tgrapperon) [inline Tests/ComposableArchitectureTests/StoreTests.swift:468]:
```suggestion
            environment: expectation
```

**Review by tgrapperon — CHANGES_REQUESTED** [review]
I think we can make the expectation the environment of the reducer.

File: Tests/ComposableArchitectureTests/StoreTests.swift (hunk 6 above)
Comment (mbrandonw) [reply Tests/ComposableArchitectureTests/StoreTests.swift]: haha, sure thing!

**Pull request merged**